File: eval/extension/tools/attempt_completion_tool.py
from eval.extension.tools.base_tool import BaseAgentTool
from eval.extension.tools.execute_command_tool import ExecuteCommandTool
from utils import format_tool_response


class AttemptCompletionTool(BaseAgentTool):
    def __init__(self, params, options):
        super().__init__(options)
        self.params = params

    async def execute(self):
        input_data = self.params.get("input", {})
        result = input_data.get("result")
        command = input_data.get("command")

        if result is None:
            return """
            Error: Missing value for required parameter 'result'. Please retry with complete response.
            An example of a good attemptCompletion tool call is:
            {
                "tool": "attempt_completion",
                "result": "result to attempt completion with"
            }
            """

        result_to_send = result
        if command:
            # Skipping the say method and directly running the command
            execute_command_params = {
                **self.params,
                "returnEmptyStringOnSuccess": True,
            }
            command_result = await ExecuteCommandTool(
                execute_command_params, self.options
            ).execute()

            if command_result:
                return command_result

            result_to_send = ""

        # Skipping the ask method, assuming approval
        return ""

    def format_tool_response(self, text, images=None):
        """Simulates formatting a tool response."""
        return format_tool_response(text, images)

File: extension/src/agent/v1/tools/attempt-completion.tool.ts
import { ToolResponse } from "../types"
import { formatToolResponse } from "../utils"
import { AgentToolOptions, AgentToolParams } from "./types"
import { BaseAgentTool } from "./base-agent.tool"
import { ExecuteCommandTool } from "./execute-command.tool"

export class AttemptCompletionTool extends BaseAgentTool {
	protected params: AgentToolParams

	constructor(params: AgentToolParams, options: AgentToolOptions) {
		super(options)
		this.params = params
	}

	async execute(): Promise<ToolResponse> {
		const { input, ask, say } = this.params
		const { result, command } = input

		if (result === undefined) {
			await say(
				"error",
				"Claude tried to use attempt_completion without value for required parameter 'result'. Retrying..."
			)
			return `Error: Missing value for required parameter 'result'. Please retry with complete response.
			An example of a good attemptCompletion tool call is:
			{
				"tool": "attempt_completion",
				"result": "result to attempt completion with"
			}
			`
		}

		let resultToSend = result
		if (command) {
			await ask("tool", {
				tool: {
					tool: "attempt_completion",
					result: result,
					command: command,
					status: "approved",
				},
			})

			const executeCommandParams = {
				...this.params,
				returnEmptyStringOnSuccess: true,
			}
			const commandResult = await new ExecuteCommandTool(executeCommandParams, this.options).execute()

			if (commandResult) {
				return commandResult
			}
			resultToSend = ""
		}

		const { response, text, images } = await ask("tool", {
			tool: {
				tool: "attempt_completion",
				result: resultToSend,
				status: "approved",
			},
		})
		if (response === "yesButtonTapped") {
			return ""
		}

		await say("user_feedback", text ?? "", images)
		return formatToolResponse(
			`The user is not pleased with the results. Use the feedback they provided to successfully complete the task, and then attempt completion again.\n<feedback>\n${text}\n</feedback>`,
			images
		)
	}
}
